### PR TITLE
[SPARK-13918] [SQL] Merge SortMergeJoin and SortMergerOuterJoin

### PR DESCRIPTION
**Pull request by davies**
## What changes were proposed in this pull request?

This PR just move some code from SortMergeOuterJoin into SortMergeJoin.

This is for support codegen for outer join.
## How was this patch tested?

existing tests.

### CODE REVIEW

**Comment by davies** [comment]
cc @rxin

**Comment by rxin** [comment]
cc @sameeragarwal for review

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoin.scala [hunk 1]
```diff
@@ -570,3 +659,307 @@ private[joins] class SortMergeJoinScanner(
...
+    val r = advanceBufferUntilBoundConditionSatisfied() || advanceStream()
+    if (r) numOutputRows += 1
+    r
+  }
+
+  override def getRow: InternalRow = resultProj(joinedRow)
+}
+
+private class SortMergeFullOuterJoinScanner(
+  leftKeyGenerator: Projection,
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoin.scala]: 4 sapce indent here

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoin.scala [hunk 2]
```diff
@@ -570,3 +659,307 @@ private[joins] class SortMergeJoinScanner(
     } while (bufferedRow != null && keyOrdering.compare(streamedRowKey, bufferedRowKey) == 0)
   }
 }
+
+/**
+ * An iterator for outputting rows in left outer join.
+ */
+private class LeftOuterIterator(
+  smjScanner: SortMergeJoinScanner,
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoin.scala]: 4 space here

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoin.scala [hunk 3]
```diff
@@ -570,3 +659,307 @@ private[joins] class SortMergeJoinScanner(
...
+
+  protected override def setStreamSideOutput(row: InternalRow): Unit = joinedRow.withLeft(row)
+  protected override def setBufferedSideOutput(row: InternalRow): Unit = joinedRow.withRight(row)
+}
+
+/**
+ * An iterator for outputting rows in right outer join.
+ */
+private class RightOuterIterator(
+  smjScanner: SortMergeJoinScanner,
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoin.scala]: 4 space here too

**Comment by rxin** [comment]
Actually this is a straightforward change. Merging in master.

**Comment by rxin** [comment]
Can you fix all the indenting issues in your next pr?

**Comment by davies** [comment]
Will do that when implement codegen support.

**Pull request closed without merging**